nemo/collections/multimodal/speech_llm/models/modular_models.py

```diff
@@ -345,6 +347,146 @@ def _get_text_embeddings(self, text_tokens, position_ids):
             text_embeddings = text_embeddings + position_embeddings
         return text_embeddings.transpose(0, 1)
 
+    def prepare_llm_input_conv(self, audio_batch):
+        """Prepare input for the LLM."""
+
+        # [b, t, c]
+        encoded, encoded_len = self.perception(
+            input_signal=audio_batch['audio_signal'],
+            input_signal_length=audio_batch['audio_signal_length'],
+            processed_signal=None,
+            processed_signal_length=None,
+        )
+
+        input_ids, input_length, loss_mask, audio_locator_ids = (
+            audio_batch['input_ids'],  # this includes bos and eos
+            audio_batch['input_id_lengths'],
+            audio_batch['loss_mask'],  # this includes bos and eos
+            audio_batch['audio_locator_ids'],
+        )
+
+        encoder_input, encoder_length, labels, loss_mask, attention_mask, position_ids = (
+            self.inject_perception_input_conv(
+                encoded=encoded,
+                encoded_len=encoded_len,
+                input_ids=input_ids,
+                input_length=input_length,
+                loss_mask=loss_mask,
+                audio_locator_ids=audio_locator_ids,
+                remove_bos_or_eos=True,
+            )
+        )
+
+        return encoder_input, attention_mask, labels, loss_mask, encoder_length
+
+    def inject_perception_input_conv(
```

Comment:
I didn't review this part in-depth; however, I think it would be best if you factored it out as a separate utility function and added unit test coverage for it for no-audio, single-audio, and multi-audio cases.